cli/arguments/reference.go

```diff
@@ -69,20 +73,39 @@ func ParseReference(arg string, parseArch bool) (*Reference, error) {
 		ret.Version = toks[1]
 	}
 
-	if parseArch {
-		toks = strings.Split(ret.PackageName, ":")
-		if len(toks) != 2 {
-			return nil, fmt.Errorf(tr("invalid item %s"), arg)
-		}
-		if toks[0] == "" {
-			return nil, fmt.Errorf(tr("invalid empty core name '%s'"), arg)
+	toks = strings.Split(ret.PackageName, ":")
+	if len(toks) != 2 {
+		return nil, fmt.Errorf(tr("invalid item %s"), arg)
+	}
+	if toks[0] == "" {
+		return nil, fmt.Errorf(tr("invalid empty core name '%s'"), arg)
+	}
+	ret.PackageName = toks[0]
+
+	// try to search in the packages, this way ParseReference can be case insensitive
+	// note that if the search fail for some reason the ret.Packagename does not get overwritten
+	var targetPackage *cores.Package
+	if inst != nil { // do not perform the search if the instance is not defined
+		pm := commands.GetPackageManager(inst.Id)
+		for _, targetPackage = range pm.Packages {
```

Comment:
We should not access the `instance` field directly (like `pm.Packages`) but we must use the RPC-mapped functions to obtain the data. In this case, I think that `core.PlatformSearch` in the `commands/core` 

```
core.PlatformSearch(&rpc.PlatformSearchRequest{
		Instance:    inst,
		SearchArgs:  arguments,
		AllVersions: allVersions,
	})
```

should return the info we need.
I'm wondering if we can make private the `Packages` field in PackageManager to avoid accidentally bypassing the gRPC API from the command-line client.

Reply:
Actually I did use the `core.PlatformSearch` function at first, then @silvanocerza suggested not to use it because the function does other stuff and can be optimized a little.

Reply:
What about `GetPlatformList` then:

```
// GetPlatforms returns a list of installed platforms, optionally filtered by
// those requiring an update.
func GetPlatforms(req *rpc.PlatformListRequest) ([]*rpc.Platform, error) {
```

Reply:
Using `GetPlatforms` is not efficient, it will always iterate all packages and all their platforms much like `PlatformSearch`.

In this case we need only find one platform, if we don't want to store packages and platforms using lower case keys this is the most efficient way.

Reply:
So what should we do about the implementation?

Reply:
We carefully avoided accessing directly the `cores` package from the `cli` package (but instead we used the `command` package to do everything). This is required to keep a feature parity between the CLI and the gRPC interface (since both use the `commands` package you can do exactly the same things using the CLI and the gRPC).
If we access directly the `Packages` field we are introducing a "feature" in the CLI that is not present in the gRPC (eventually leading to code duplication/feature incoherence).

> Using GetPlatforms is not efficient, it will always iterate all packages and all their platforms much like PlatformSearch.

I don't think any user will ever be able to see this optimization since the bulk of the operations is the loading of the platforms at init time (that sometimes takes up to a second), the few milliseconds spent iterating a bunch of platforms is negligible.

Anyway, in the platform specification, the platform name **is** case sensitive, so we cannot stop at the first match, because there could be more than one platform matching: in that case the CLI should output an error and ask the user to disambiguate.

Reply:
> We carefully avoided accessing directly the `cores` package from the `cli` package (but instead we used the `command` package to do everything). This is required to keep a feature parity between the CLI and the gRPC interface (since both use the `commands` package you can do exactly the same things using the CLI and the gRPC).
If we access directly the `Packages` field we are introducing a "feature" in the CLI that is not present in the gRPC (eventually leading to code duplication/feature incoherence).

You're right, didn't think about this. :thinking: 

> Anyway, in the platform specification, the platform name **is** case sensitive, so we cannot stop at the first match, because there could be more than one platform matching: in that case the CLI should output an error and ask the user to disambiguate.

This concerns me though, if platform names are case sensitive and we make commands work without sensitivity I think it may cause issues further in the future. Is it actually worth making this change? :confused:

Reply:
Ok I'm trying to use the `core.GetPlatform()` approach and see what happens